src/Sylius/Bundle/ApiBundle/Validator/GatewayConfig/GatewayConfigValidator.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Validator\GatewayConfig;
+
+use Sylius\Bundle\ApiBundle\Validator\Constraints\GatewayConfig;
+use Sylius\Bundle\PayumBundle\Model\GatewayConfigInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+use Symfony\Component\Validator\Exception\UnexpectedValueException;
+
+final class GatewayConfigValidator extends ConstraintValidator
+{
+    public const OFFLINE_GATEWAY_FACTORY = 'offline';
+
+    /** @param array<string, string> $factoryNames */
+    public function __construct(private array $factoryNames)
+    {
+    }
+
+    /** @param GatewayConfigInterface $value */
+    public function validate($value, Constraint $constraint): void
+    {
+        if (!$constraint instanceof GatewayConfig) {
+            throw new UnexpectedTypeException($constraint, GatewayConfig::class);
+        }
+
+        if (!$value instanceof GatewayConfigInterface) {
+            throw new UnexpectedValueException($value, GatewayConfigInterface::class);
+        }
+
+        /** @var string|null $factoryName */
+        $factoryName = $value->getFactoryName();
+
+        if ($factoryName === self::OFFLINE_GATEWAY_FACTORY || $factoryName === null) {
+            return;
+        }
+
+        if (!in_array($factoryName, array_keys($this->factoryNames), true)) {
+            $this->context->buildViolation($constraint->invalidGatewayFactory)
+                ->atPath('factoryName')
+                ->setParameter('{{ available_factories }}', implode(', ', array_keys($this->factoryNames)))
+                ->addViolation()
+            ;
+
+            return;
+        }
+
+        /** @var string[] $groups */
+        $groups = array_merge($constraint->groups, [$factoryName]);
```

Comment:
Usage of `$factoryName` as validation group in plugin context feels like some kind of rule breakage/design leakage, which I cannot properly name.

`sylius/sylius` source codes shouldn't change it's  behavior/react on plugin with payum factory just due fact it's installed/active.

Reply:
What do you mean by changing behavior?

Reply:
> What do you mean by changing behavior?

It's more of conceptual issue -- validates `GatewayConfig` with external validation group without any prior configuration, knowledge or control over this.

Here is case,
I've developed new payment plugin with new payum factory and installing it into sylius now
And sylius suddenly validates `GatewayConfig` with my `acme_new_plugin` group.

Problem,
- As plugin author, I didn't configure or ask for this behavior
- I'm unaware it even exists
- sylius/sylius changes it's behavior (validates `GatewayConfig` with external group)
- As developer I cannot possibly expect something like this to happen -- such practice is not common or even exists in SF/sylius.
- As plugin author I want to be in control what validates when and not be forced to use it or look for workarounds.

Resume: If it needs to become part of regular plugin flow, then we should provide compiler pass or something similar.

Reply:
Okay, I understand that there could be an issue. 
I attempted to use a different custom factory, and if it is properly specified as shown [here](https://docs.sylius.com/en/1.12/cookbook/payments/custom-payment-gateway.html), it will be automatically added to the `sylius.gateway_factories` parameter through the `RegisterGatewayConfigTypePass`. It will then pass the condition as outlined [here](https://github.com/Sylius/Sylius/pull/15549/files#diff-680393fd06307f25dfabb502b733d7c92605b3874ff713e5ac29d57b6523732fR50). Validation will not cause any issues if the group does not exist, so the object will be updated/created. Additionally, it's important to note that these constraints only work for the API; the UI is validated through constraints on the forms.
I think in this case we have control over this, as If we have a properly specified gateway that's different, it won't be validated unless we add the constraints

Reply:
@Wojdylak found this solution for the same issue in CatalogPromotion codebase https://github.com/Sylius/Sylius/blob/1.13/src/Sylius/Bundle/PromotionBundle/Validator/CatalogPromotionActionValidator.php#L53C1-L53C1

What do you think?

Reply:
https://github.com/Sylius/Sylius/pull/15549#discussion_r1401763261